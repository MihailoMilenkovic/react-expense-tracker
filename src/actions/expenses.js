import uuid from "uuid";
import database from "../firebase/firebase";
//without a database:
//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes

//with a database:
//component calls action generator
//action generator returns function
//component dispatches function
//function runs (can dispatch other actions and do other things)

export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database
      .ref("expenses")
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          })
        );
      });
  };
};

//remove expense
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

//edit expense
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});
