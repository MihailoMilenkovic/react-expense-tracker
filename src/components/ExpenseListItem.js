import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
//desc, amount. createdAt
const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <h3>{description}</h3>
    <p>
      {numeral(amount / 100).format("$0,0.00")}
      -
      {moment(createdAt).format("Do MMMM YYYY")}
    </p>
    <Link to={`/edit/${id}`}>edit expense</Link>
  </div>
);

export default ExpenseListItem;
