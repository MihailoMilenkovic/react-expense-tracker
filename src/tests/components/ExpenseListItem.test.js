import React from "react";
import {shallow} from "enzyme";
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("should render expense item correctly",()=>{
    const wrapper=shallow(<ExpenseListItem key={expenses[0].id} {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});