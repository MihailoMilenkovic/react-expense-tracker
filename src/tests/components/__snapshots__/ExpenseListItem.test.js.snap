// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense item correctly 1`] = `
<div>
  <h3>
    Gum
  </h3>
  <p>
    $1.95
    -
    1st January 1970
  </p>
  <Link
    to="/edit/1"
  >
    edit expense
  </Link>
</div>
`;
